### PR TITLE
Rework async parallel pattern 

### PR DESCRIPTION
**Pull request by reble**
This PR extracts changes from #78 to device backends that effect existing layout. So that all existing async pattern return a future<void> which holds a sycl::event and can be reused by Async API.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h [hunk 1]
```diff
@@ -1305,37 +1305,37 @@ __pattern_inplace_merge(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator
 //------------------------------------------------------------------------
 // sort
 //------------------------------------------------------------------------
-template <typename _ExecutionPolicy, typename _Iterator, typename _Compare>
+template <typename _IsSync = ::std::true_type, typename _ExecutionPolicy, typename _Iterator, typename _Compare>
```
[the remark this answers is not in the document]
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1308]: I reused the naming of `__pattern_walk2` which has a kind-of async mode. If we rename the template parameter here, we should do it for both pattern.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 2]
```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
+//-----------------------------------------------------------------------
+// future and helper classes for async pattern/algorithm
+//-----------------------------------------------------------------------
+
+struct __tmp_base
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:607]: What is this empty base class for?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 3]
```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
...
+{
+};
+
+template <typename... Ts>
+struct __TempObjs : public __tmp_base
+{
+    ::std::tuple<Ts&...> __my_tmps;
+    __TempObjs(Ts&... __t) : __my_tmps(::std::forward_as_tuple(__t...)) {}
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:615]: Why `forward_as_tuple` in the context where it's nothing to forward?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 4]
```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
...
+};
+
+class __future_base
+{
+    using event = sycl::event;
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:620]: Do you really need this alias?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 5]
```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
...
+#if !ONEDPL_ALLOW_DEFERRED_WAITING
+        __my_event.wait();
+#endif
+    }
+    operator event() const { return event(__my_event); }
+};
+
+// TODO: Extend to support value type and sycl iterator.
+template <typename T>
+class __future : public __future_base
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:642]: When do we want to support value_type? in the upcoming release?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 6]
```diff
@@ -600,6 +600,66 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
...
+};
+
+template <>
+class __future<void> : public __future_base
+{
+    __tmp_base __tmps;
+
+  public:
+    template <typename... _Ts>
+    __future(sycl::event __e, _Ts... __t) : __future_base(__e), __tmps(__TempObjs<_Ts...>{__t...})
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:653]: I don't understand what's going on. You create the `__TempObjs<_Ts...>{__t...}` that immediately dies and you just keep `__tmp_base` member as empty class. What is the value?

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmiesmith) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1308]: Ok. Let's keep it consistent with the other components.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 3 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: I've added a check to the caller. Should be sufficient.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 6 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: Good catch. Design was broken, should be fixed now.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 5 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:638]: It's only needed for non-void futures. I was planning to keep them separate for the next release.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h [hunk 7]
```diff
@@ -1305,37 +1305,37 @@ __pattern_inplace_merge(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator
...
 oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, void>
 __pattern_sort(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last, _Compare __comp,
                /*vector=*/::std::true_type, /*parallel=*/::std::true_type, /*is_move_constructible=*/::std::true_type)
 {
     if (__last - __first < 2)
         return;
 
-    __par_backend_hetero::__parallel_stable_sort(
+    auto __future_obj = __par_backend_hetero::__parallel_stable_sort(
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1316]: I realized the issue that if we don't need to wait (basically it means `_IsSync` is `std::false_type` then we need to return that future to the user

Comment (reble) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1322]: If we change the return type at this level, I see compilation errors. Type mismatch vs. void function is returning something. Async algorithm can call parallel sort pattern directly.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 8]
```diff
@@ -600,6 +600,79 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
...
+//-----------------------------------------------------------------------
+
+// empty base class for type erasure
+struct __tmp_base
+{
+};
+
+// derived class to keep temporaries (e.g. buffer) alive
+template <typename... Ts>
+struct __TempObjs : public __tmp_base
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:614]: All the types in the patch a named in snake_case. Only this one is named with PascalCase.

More important: the name doesn't reflect the purpose of the type. I would name it something like `lifetime_extender` or `lifetime_keeper` or `objects_keeper`. Please choose whatever you like or propose yours.

Please rename `tmp_base` correspondently.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h [hunk 9]
```diff
@@ -600,6 +600,79 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
+//-----------------------------------------------------------------------
+// future and helper classes for async pattern/algorithm
+//-----------------------------------------------------------------------
+
+// empty base class for type erasure
+struct __tmp_base
```
Comment (rarutyun) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h:608]: Currently we have memory leak. You store the base class in the `std::unique_ptr`. But you don't have virtual destructor. So, for all `Ts...` in `TempObjs` destructors are not called

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h (hunk 8 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h]: Makes sense. Done.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h (hunk 7 above)
Comment (reble) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h]: I removed the template parameter to switch between sync and async behavior at that level. Async sort variants will call `parallel_stable_sort` pattern directly.

**Review by timmiesmith — APPROVED** [review]
This looks good to me.

**Review by timmiesmith — APPROVED** [review]

**Review by timmiesmith — DISMISSED** [review]

**Comment by capatober** [comment]
@MikeDvorskiy, could you please review changes?

**Review by timmiesmith — APPROVED** [review]

**Comment by capatober** [comment]
@reble, please, do not forget to resolve discussions with comments even if the PR is merged.

**Pull request merged**